doc/arch/011-screen-navigation-v2.md

```diff
@@ -0,0 +1,263 @@
+
+## Goals
+We need a new navigation framework, one that is better suited for our needs. It needs to fulfill the following requirements:
+
+1. Have in-built support for modals (bottom sheets and dialogs). Ideally, there should be minimal difference between using a modal vs a normal screen
+2. Support the Android Lifecycle component
+3. Support opening screens and returning a result without requiring workarounds
+4. Support overriding back presses at the screen/modal level
+5. Support layout editor previews in the IDE
+6. Be easy to incrementally migrate from our current navigation architecture without requiring a lot of upfront work
+
+## Solution
+We will build a new navigation framework based on [Fragments](https://developer.android.com/guide/components/fragments), inspired by [simple-stack](https://github.com/Zhuinden/simple-stack), and designed to work with our screen architecture (See [ADR #008](https://github.com/simpledotorg/simple-android/blob/master/doc/arch/008-screen-architecture-v3.md)).
+
+A high level description of the framework and its core components are described below:
+
+### Framework
+The framework overall will have three primary components:
+
+#### `ScreenKey`
+This will be an abstract class that will need to be inherited by all classes that represent individual screens, similar to the current implementation. This will need to be `Parcelable` so that we can save and restore the backstack when the activity state restoration mechanism is triggered.
+
+#### `History`
+This is a class that represents the overall backstack. This will encapsulate a list of `ScreenKey` instances and expose some helper methods to manipulate the backstack. This will also need to be `Parcelable` so that we can save and restore it as necessary.
+
+#### `Router`
+This is the class that will be used externally to push and pop screens from the backstack. This will manage converting the `History` into fragment transactions that will then be applied in order to update the UI.
+
+### Usage
+Instantiating a `Router` is as simple as
+```kotlin
+val router = Router(
+  FirstScreen.Key(), // Initial screen
+  supportFragmentManager,
+  R.id.content // Reference to the fragment container viewgroup
+)
+```
+
+Pushing a new screen onto the stack can be performed using the `push()` method.
+```kotlin
+nextScreenButton.setOnClickListener { router.push(SecondScreen.Key()) }
+```
+
+Going to the previous screen is done using the `pop()` method.
+```kotlin
+previousScreenButton.setOnClickListener { router.pop() }
+```
+
+#### Building a screen
+Implementing a screen is quite simple, it's a standard fragment.
+
+```kotlin
+class FirstScreen : Fragment(R.layout.screen_navigation_first) {
```

Comment:
Good point, I'll update the ADR.

Reply:
Never mind, I have decided not to make this change here because this router can be used completely independently of the navigation framework by just using fragments directly. This is more of a follow up ADR to [008](../doc/arch/008-screen-architecure-v3.md) which talks about how to integrate Mobius into our screens.

I'll raise a separate ADR for that.